rest/model/codegen.go

```diff
@@ -63,84 +63,113 @@ type modelConversionInfo struct {
 // ModelMapping maps schema type names to their respective DB model
 type ModelMapping map[string]string
 
-func Codegen(schema string, config ModelMapping) ([]byte, error) {
+func Codegen(schema string, config ModelMapping) ([]byte, []byte, error) {
 	source := ast.Source{
 		Input: schema,
 	}
 	fileTemplate, err := getTemplate(fileTemplatePath)
 	if err != nil {
-		return nil, errors.Wrap(err, "error loading file template")
+		return nil, nil, errors.Wrap(err, "error loading file template")
 	}
 	structTemplate, err := getTemplate(structTemplatePath)
 	if err != nil {
-		return nil, errors.Wrap(err, "error loading struct template")
+		return nil, nil, errors.Wrap(err, "error loading struct template")
 	}
 	fieldTemplate, err := getTemplate(fieldTemplatePath)
 	if err != nil {
-		return nil, errors.Wrap(err, "error loading field template")
+		return nil, nil, errors.Wrap(err, "error loading field template")
 	}
 	parsedAst, parseErr := gqlparser.LoadSchema(&source)
 	if parseErr != nil {
-		return nil, errors.Wrap(parseErr, "error parsing schema")
+		return nil, nil, errors.Wrap(parseErr, "error parsing schema")
 	}
 	catcher := grip.NewBasicCatcher()
 	structs := ""
-	conversionCode := ""
+	serviceCode := ""
+	generatedConversions := map[string]string{}
 
 	typeNames := []string{}
 	for key := range parsedAst.Types {
 		typeNames = append(typeNames, key)
 	}
 	sort.Strings(typeNames)
 	for _, typeName := range typeNames {
-		dbModel, shouldConvert := config[typeName]
-		if !shouldConvert {
+		gqlType := parsedAst.Types[typeName]
+		if gqlType == nil {
+			catcher.Errorf("unable to find type '%s'", typeName)
 			continue
 		}
-		gqlType := parsedAst.Types[typeName]
 		if gqlType.BuiltIn {
 			continue
 		}
 		if gqlType.Kind == ast.Scalar {
 			continue
 		}
-		fields := ""
-		extractedFields := extractedFields{}
-		for _, field := range gqlType.Fields {
-			fieldInfo := getFieldInfo(field)
-			extractedFields[field.Name] = fieldInfo
-			var fieldData string
-			fieldData, err = output(fieldTemplate, fieldInfo)
-			if err != nil {
-				catcher.Add(err)
-				continue
-			}
-			fields += fieldData
+		dbModel, shouldConvert := config[typeName]
+		if !shouldConvert {
+			continue
 		}
-		var structData string
-		structData, err = output(structTemplate, structInfo{Name: typeName, Fields: fields})
+
+		structData, code, err := generateForStruct(fieldTemplate, structTemplate, typeName, dbModel, *gqlType, generatedConversions)
 		if err != nil {
 			catcher.Add(err)
 			continue
 		}
 		structs += structData
+		serviceCode += code
+	}
+	modelFile, err := output(fileTemplate, fileInfo{Package: "model", Structs: structs, Code: serviceCode})
+	catcher.Add(err)
+	formattedModelCode, err := goimports(modelFile)
+	catcher.Add(err)
```

Comment:
Can this error be wrapped? The error might not be very pretty

Reply:
done